### PR TITLE
Changing color editors to accept #AARRGGBB or #RRGGBB html codes as i…

### PR DESCRIPTION
**Pull request by josephcrowell**
…nput instead of 3 spin edits.

TODO: Change display when out of editor to show similar values and mask input to only allow proper values.

### CODE REVIEW

**Comment by neomonkeus** [comment]
@jonwd7 Can you review

**Comment by hexabits** [comment]
For whatever reason didn't see this until now.  I'll take a look.

Right off the bat, I notice a lot of bad whitespace changes.   If you're going to make pull requests you should edit the files in whatever mode that particular file uses.  Is this case it seems @smokex is using spaces instead of tabs.

~~Also, I'm not sure what, if anything, I've changed in the 2.0 branch that may collide with this.  I was expecting the 2.0 branch to be merged to develop by now so that we're all working off the same base.~~ I see you were working of dev/2.0  :) 

Finally,  it'd be nice if the pull request included some details about why this change is being made.  My initial reaction is to reject this change because there is no argument or explanation made for its inclusion.  How is the data stored in the actual NIF?  If it's 3 or 4 consecutive integers,  then I think we ~~should~~must keep the 3/4 spin edits.  Does it remove the spin edits or supplement them?  For example in the rotation editor, there is a button to switch between two ways of editing rotations.

I've actually had someone complain to me about my inclusion of the hex editor in the Settings panel for the wireframe/background/etc colors.  Their argument was essentially that they were not a designer and did not understand hex codes.  I said that in this particular case, I am also giving a color wheel, so the hex edit boxes are specifically for people who want an exact color.  For non-designers the color wheel is sufficient and so they don't need the separate R/G/B/A spin edits.  

However, for the actual NIF data the story is very different.  The Block Details pane is supposed to be a tabular representation of the file data with no abstractions (within reason).  If the color rows are represented in the file as 3 or 4 consecutive integers, then we must keep that way of editing the data.  This is why supplemental widgets have been added to these particular rows, like the Color Wheel icon.  If the AARRGGBB etc format should be entered anywhere, it should be in the Color Wheel widget, not the actual editor in the table. (Or a button to switch between the two editing modes as with Rotations)

I'll look into how the colors are actually stored later. 

---

tl;dr - I have nothing against hex color codes, but I know some people do.  Also, if the color is stored as consecutive integers in the actual raw NIF data, we need to maintain the ability to view/edit it that way.  Editing as hex should be a supplemental addition.

**Comment by josephcrowell** [comment]
This change is against the 2.0 branch. I'm developing on it.

[@I've](https://sourceforge.net/u/smokexyz/profile/) had about 15 users that have asked for the ability to copy and paste color values quickly to make the editing of vertex colors a bit less of a pain. They specifically ask for the ability to copy/paste css style color codes.

We could have the spin edits with a remembered setting to switch to this mode. If it opens with the spin edits with the requirement to switch to these editors every time, it defeats the purpose.

The colors are stored as floats.

I was using the Qt default whitespace because I was informed that whitespace wasn't a big deal here. After seeing that these particular files use 8 character tab spacing, I could change my Qt settings but it would be better to include a Qt user template to set the white spacing so that everyone gets the correct settings on initial clone.

**Comment by josephcrowell** [comment]
jondw7:

> The Block Details pane is supposed to be a tabular representation of the file data with no 
> abstractions (within reason). If the color rows are represented in the file as 3 or 4 consecutive 
> integers, then we must keep that way of editing the data. 

I'm also rather surprised by this. You're saying the end user is supposed to have an application that is functionally harder to use so that they can see underlying data structures that are almost certainly useless to them?

All of the existing editors provide a level of abstraction over the existing data structures including those spin edits which are in fact a common complaint in this application.

**Comment by hexabits** [comment]
> The colors are stored as floats.

Ah, the use of the term "spin edit" confused me as I assumed then that the color editor must be integer based (RGB 0-255).  I didn't look through it enough to notice the editor is using floats.  (Generally in the code spin edit refers to the integer spin edit.)

So, just replace every time I said int/integer with "float", and my argument remains the same.  The data is stored in floats so we need to be able to display/access/modify it as such.

> You're saying the end user is supposed to have an application that is functionally harder to use so that they can see underlying data structures that are almost certainly useless to them?

It's primarily about data integrity.  It's also called "Block Details" for a reason.  It's a very scary area and it's not really meant for people who wouldn't consider themselves "power users".   My 2.0 plans consist of hiding away a lot of the "scariness" my bringing in another pane which is meant to display the data in a more user-friendly manner.  The "Block Details" pane is intended to be the fully tabular representation of a block's data, including the bits users should _never_ touch.  It was never promised to be user-friendly, it's just that no one has bothered providing a user-friendly alternative yet.

Anyway, back to the data integrity.  So the colors are stored as consecutive floats.  Here is an actual color listed in a NIF:

R 0.478431
G 0.423529
B 0.372459
A 1

this is reported as  #7a6c5fff. 

Currently, NifSkope is actually doing something **very bad**, which I intend on fixing.  When looking into this I noticed that 3 decimal places is simply not enough for showing the colors.  When you double click the color and then click away the file acts as if it's been modified, even if you don't change the spin values.  This is because it's actually truncating the floats to 3 places.  In the case of this specific color it actually changes from #7a6c5fff to #796c5fff. 

Now, I'm going to guess that the creation process Bethesda uses for NIFs involves a color picker which produces only hex codes.  But to verify this, I'm going to have to go through lots of colors and see if any of the floats are not perfectly represented by an integer divided by 255.  In the above example R = 122, G = 108, B = 95.    However, that's only Bethesda's NIFs.  I can't vouch for any others at the moment.  The format itself specifies the colors as floats, so colors between 122/255 (0.478431) and 123/255 (0.482353) are perfectly valid as far as I'm aware, so NifSkope needs to be able to both display and reproduce this.

Realistically, people editing vertex colors are also not going to be using NifSkope.  Yet.  After the new renderer is written I do eventually want actual vertex painting and such, but that's a long way away. 

> I've had about 15 users that have asked for the ability to copy and paste color values quickly to make the editing of vertex colors a bit less of a pain.

Yes, and that's why **generic** copy/pasting of _any_ NifValue is on the roadmap for 2.0.  As in, you can click on a cell in the Value column, hit Ctrl+C, and then Ctrl+V over another cell.   I've also had many people request copy/pasting for the same reasons, but opting for the generic solution is much better.

If you want I can add you to NifTools' Trello board which houses the NifSkope roadmap.   I'm also curious, who are the 15 users?  I don't even have that many people I have access to for input on NifSkope (the Unofficial Patch team).  I'd like to know what your people have to say.  They may have a lot of the same input as my people,  or a lot of conflicting input.  

> I was using the Qt default whitespace because I was informed that whitespace wasn't a big deal here.

I started but never finished the style guide for NifSkope: https://github.com/niftools/nifskope/wiki/Coding-Style-Guide

... because writing such things is soul-sucking.  :)   I didn't even get to the part where I mention tabs vs. spaces there.

Generally, you just match the source file when it comes to tabs/spaces or you get a lot of whitespace-only edits as in your commit.  If you notice lines where nothing but the whitespace changes, those absolutely cannot be committed at the same time as non-whitespace commits.  Basically the coding "best practice" is to always separate your whitespace- and code- changing commits. 

> We could have the spin edits with a remembered setting to switch to this mode. If it opens with the spin edits with the requirement to switch to these editors every time, it defeats the purpose.

I agree, which is why it's probably best to have it as a user setting in the Settings panel.  "Edit colors as:" and have the choice of float/int/hex.  But again, I'm not sure if it's necessary if your use case is only about simplifying copy/pastes because generic copy/pastes for NifValues is planned.

**Comment by josephcrowell** [comment]
The 15 users are users at Dark Creations. Arthmoor and zilav pop in there from time to time and I think Neo has been there before. I guess you recognize me from afkmods?

I've been a bit busy lately leading one of the Beyond Skyrim provinces but maybe I'll get a bit of time soon and can help get this color copy/paste sorted properly. :)

**Comment by hexabits** [comment]
@smokex @neomonkeus   After thinking about it for a bit, I think I'm fine with the default being hex, as long as there is a button to switch to RGB float and RGB int, as with the Rotations editor.

This is assuming that 99.9% of the floats in all NIF files for Color3/Color4 value types correspond with the number `x/255`, where x ranges from `{0,255}` **.  In this case, showing them as hex is actually _more_ accurate than current builds of NifSkope, where the floats are incorrectly truncated to 3 digits.

The editor could be programmed to detect floats that do NOT correspond with those 256 floating point values, within a certain error threshold (due to floating point rounding issues).  If one of the floats in the RGB(A) values are not within the threshold, the editor can automatically switch to the float editor. (As long as the precision is fixed in the current float editor)

---

**:  I still have yet to audit a wide range of NIF files to determine if this is the case.

**Comment by josephcrowell** [comment]
I'm also going to create a .pro.shared file which will specify the whitespace settings for the project files. 

http://doc.qt.io/qtcreator/creator-sharing-project-settings.html

It will look similar to this: https://sourceforge.net/p/quazaa/gitcode/ci/master/tree/Quazaa.pro.shared

**Comment by hexabits** [comment]
Yeah, that'd be a nice inclusion, but personally I use Visual Studio for all my actual editing.  I just use Qt Creator for building MinGW builds.
